compute_tools/src/monitor.rs

```diff
@@ -6,197 +6,296 @@ use chrono::{DateTime, Utc};
-    } else {
-        info!("starting activity monitor for {}", connstr);
+struct ComputeMonitor {
+    compute: Arc<ComputeNode>,
+
+    // The moment when Postgres had some activity,
+    // that should prevent compute from being suspended.
+    last_active: Option<DateTime<Utc>>,
+
+    // The moment when we last tried to check Postgres.
+    last_checked: DateTime<Utc>,
+    // The last moment we did a successful Postgres check.
+    last_up: DateTime<Utc>,
+
+    // These two attributes are purely for statistics change tracking,
+    // they can be outdated.
+    active_time: Option<f64>,
+    sessions: Option<i64>,
+
+    // Use experimental statistics-based activity monitor. It's no longer
+    // 'experimental' per se, as it's enabled for everyone, but we still
+    // keep the flag as an option to turn it off in some cases if it will
+    // misbehave.
+    experimental: bool,
+}
+
+impl ComputeMonitor {
+    fn report_down(&mut self) {
+        let now = Utc::now();
+
+        // Calculate and report current downtime
+        // (since the last time Postgres was up)
+        let downtime = now.signed_duration_since(self.last_up);
+        PG_CURR_DOWNTIME_MS.set(downtime.num_milliseconds() as f64);
+
+        // Calculate and update total downtime
+        // (cumulative duration of Postgres downtime in ms)
+        let inc = now
+            .signed_duration_since(self.last_checked)
+            .num_milliseconds();
+        PG_TOTAL_DOWNTIME_MS.inc_by(inc as u64);
     }
 
-    loop {
-        // We use `continue` a lot, so it's more convenient to sleep at the top of the loop.
-        // But skip the first sleep, so we can connect to Postgres immediately.
-        if sleep {
-            // Should be outside of the mutex lock to allow others to read while we sleep.
-            thread::sleep(MONITOR_CHECK_INTERVAL);
-        } else {
-            sleep = true;
-        }
+    fn report_up(&mut self) {
+        self.last_up = Utc::now();
+        PG_CURR_DOWNTIME_MS.set(0.0);
+    }
 
-        match &mut client {
-            Ok(cli) => {
-                if cli.is_closed() {
-                    info!("connection to Postgres is closed, trying to reconnect");
+    fn downtime_info(&self) -> String {
+        format!(
+            "total_ms: {}, current_ms: {}, last_up: {}",
+            PG_TOTAL_DOWNTIME_MS.get(),
+            PG_CURR_DOWNTIME_MS.get(),
+            self.last_up
+        )
+    }
 
-                    // Connection is closed, reconnect and try again.
-                    client = conf.connect(NoTls);
-                    continue;
-                }
+    /// Spin in a loop and figure out the last activity time in the Postgres.
+    /// Then update it in the shared state. This function never errors out.
+    /// NB: the only expected panic is at `Mutex` unwrap(), all other errors
+    /// should be handled gracefully.
+    #[instrument(skip_all)]
+    pub fn run(&mut self) {
```

Comment:
It looks like this function never returns. I believe we can annotate the return type as `!`.

Reply:
Looks like it's an experimental feature https://github.com/rust-lang/rust/issues/35121

I see it's used in a couple of places, but here it fails with
```
-> % cargo check
    Checking compute_tools v0.1.0 (/Users/alexk/dev/neon/compute_tools)
error[E0658]: the `!` type is experimental
  --> compute_tools/src/monitor.rs:78:30
   |
78 |     pub fn run(&mut self) -> ! {
   |                              ^
   |
   = note: see issue #35121 <https://github.com/rust-lang/rust/issues/35121> for more information
```
for me for some reason